tee-worker/litentry/core/data-providers/src/achainable.rs

```diff
@@ -740,6 +779,69 @@ impl AchainableAccountTotalTransactions for AchainableClient {
 	}
 }
 
+pub trait AchainableUtils {
+	fn get_balance(response: serde_json::Value) -> Result<f64, Error>;
+}
+
+impl AchainableUtils for AchainableClient {
+	fn get_balance(response: serde_json::Value) -> Result<f64, Error> {
+		let display_text = response
+			.get("display")
+			.and_then(|displays| {
+				displays.as_array().map(|displays| {
+					let mut text: std::option::Option<String> = None;
+					for v in displays.iter() {
+						text = v
+							.get("text")
+							.and_then(|text| {
+								text.as_str().map(|display_text| Some(display_text.to_string()))
+							})
+							.flatten();
+					}
+					text
+				})
+			})
+			.flatten();
+		if let Some(display_text) = display_text {
+			// TODO:
+			// text field format: Balance over 0 (Balance is 588.504602529)
+			let split_text = display_text.split("Balance is ").collect::<Vec<&str>>();
+			if split_text.len() != 2 {
+				return Err(Error::AchainableError("Invalid array".to_string()))
+			}
+
+			let mut value_text = split_text[1].to_string();
+
+			// pop the last char: ")"
+			value_text.pop();
+
+			let value: f64 = value_text.parse::<f64>().unwrap_or_default();
+
+			return Ok(value)
+		}
+
+		Err(Error::AchainableError("Invalid response".to_string()))
+	}
+}
+
+pub trait HoldingAmount {
+	fn holding_amount(&mut self, addresses: Vec<String>, param: Params) -> Result<String, Error>;
+}
+impl HoldingAmount for AchainableClient {
+	fn holding_amount(&mut self, addresses: Vec<String>, param: Params) -> Result<String, Error> {
+		let mut total_balance = 0_f64;
+		addresses.iter().for_each(|address| {
+			let body = ReqBody::new(address.into(), param.clone());
+			debug!("Holding amount body: {:?}", body);
+			let balance =
+				self.post(SystemLabelReqPath::default(), &body).and_then(Self::get_balance);
+			total_balance += balance.unwrap_or_default();
```

Comment:
I think we agreed that if http request fails we should error out - otherwise the `total_balance` will be simply wrong => VC is wrong

Please check other places too